signer/core/timeout.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of go-ethereum.
+//
+// go-ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// go-ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with go-ethereum. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/internal/ethapi"
+	"github.com/ethereum/go-ethereum/log"
+	"time"
+)
+
+// The TimedExternalAPI implements ExternalAPI, but can be configured to time out after a specified interval.
+// This can be used to ensure that callers get a response within reasonable time,
+// even if the user is unresponsive.
+type TimedExternalAPI struct {
+	timeout time.Duration
+	next    ExternalAPI
+}
+
+func NewTimedExternalAPI(next ExternalAPI, timeout time.Duration) ExternalAPI {
+	return &TimedExternalAPI{timeout, next}
+}
+
+var (
+	ErrTimeout = fmt.Errorf("timeout occurred")
+)
+
+func (t *TimedExternalAPI) List(ctx context.Context) ([]common.Address, error) {
+	type response struct {
+		addr []common.Address
+		err  error
+	}
+	ch := make(chan response)
+	go func() {
+		addr, err := t.next.List(ctx)
+		ch <- response{addr, err}
+	}()
+
+	select {
+	case r := <-ch:
+		return r.addr, r.err
+	case <-time.After(t.timeout):
+		log.Info("timeout", "op", "list")
+		go func() { <-ch }()
```

Comment:
So what I'm curious about here... If I set the channel to be buffering, with a size of `1`, then the `ch <- response{addr, err}` won't block, and thus allow the goroutine to exit. So can I do that, and then skip this part where I clean out the channel? Or will that leave residue in the system?

Reply:
If all references disappear to a channel, it will be garbage collected along with all its content. You don't need to explicitly drain a channel.